### PR TITLE
[EAGLE-3698]-split-handle batch

### PR DESCRIPTION
**Pull request by phatvo9**
# What
Split part of #227, handle batch

### CODE REVIEW

File: clarifai/models/model_serving/models/inference.py [hunk 1]
```diff
@@ -10,6 +10,11 @@
 import os
 from pathlib import Path
 
+from clarifai.models.model_serving.model_config import (  # noqa # pylint: disable=unused-import
+    ModelTypes, get_model_config)
+
+#config = get_model_config("clarifai-model-type")
```
Comment (ackizilkale) [inline clarifai/models/model_serving/models/inference.py:16]: remove this line

**Review by ackizilkale — APPROVED** [review]
This is a tough PR by nature, since we have to update all these examples to handle batching. With examples comes the difficulty to maintain obviously :(

But looking good to me. Please wait for Harmit's comments also.

File: clarifai/models/model_serving/models/inference.py (hunk 1 above)
Comment (HarmitMinhas96) [reply clarifai/models/model_serving/models/inference.py:16]: It is a bit confusing that this is commented out and then replaced. Maybe leave a comment here about the behavior?

Additionally, it might be a little clearer if `"clarifai-model-type"` is replaced with something like `"MODEL_TYPE_PLACEHOLDER"`, and then the replacement code only replaces `MODEL_TYPE_PLACEHOLDER`

File: clarifai/models/model_serving/models/inference.py [hunk 2]
```diff
@@ -24,19 +29,22 @@ def __init__(self) -> None:
     #self.checkpoint_path: Path = os.path.join(self.base_path, "your checkpoint filename/path")
     #self.model: Callable = <load_your_model_here from checkpoint or folder>
 
-  #Add relevant model type decorator to the method below (see docs/model_types for ref.)
-  def get_predictions(self, input_data, **kwargs):
+  #@config.inference.wrap_func
```
Comment (HarmitMinhas96) [inline clarifai/models/model_serving/models/inference.py:32]: If you don't comment out the above, I don't think this needs to be commented out either. And then it wouldn't have to be replaced in `build_repository` as well

File: clarifai/models/model_serving/pb_model_repository.py [hunk 3]
```diff
@@ -80,22 +82,30 @@ def build_repository(self, repository_dir: Path = os.curdir):
...
+                    f"""config = get_model_config("{self.model_config.model_type}")""")
+      x = x.replace("#@config.inference.wrap_func", "@config.inference.wrap_func")
+      return x
+
+    inference_py_path = os.path.join(model_version_path, "inference.py")
+    self._module_to_file(inference, inference_py_path, inference_insert_model_type_func)
+
     # generate test.py
+    def insert_model_type_func(x):
+      return x.replace("clarifai-model-type", self.model_config.model_type)
```
Comment (HarmitMinhas96) [inline clarifai/models/model_serving/pb_model_repository.py:108]: It would be good to also update test.py to use similarly named `MODEL_TYPE_PLACEHOLDER` and then you only need one helper function for inference.py and test.py

File: clarifai/models/model_serving/models/model_types.py [hunk 4]
```diff
@@ -39,11 +39,12 @@ def parse_predictions(self, input_data: np.ndarray, *args, **kwargs):
     out_bboxes = []
     out_labels = []
     out_scores = []
-    for item in input_data:
-      preds = func(self, item, *args, **kwargs)
-      out_bboxes.append(preds.predicted_bboxes)
-      out_labels.append(preds.predicted_labels)
-      out_scores.append(preds.predicted_scores)
+    input_data = [each for each in input_data]
```
Comment (HarmitMinhas96) [inline clarifai/models/model_serving/models/model_types.py:42]: What's the purpose of this line? It looks like it's creating a local copy of input_data, but why? If it is needed, it can be simplified to `input_data[:]`, I think.

I also see this is done a few more times in the other wrapper definitions.

File: clarifai/models/model_serving/cli/deploy_cli.py
Comment (HarmitMinhas96) [inline clarifai/models/model_serving/cli/deploy_cli.py:1]: How does this user specify their batch size now? Should we add an arg for it? Or will that be in another PR?

**Review by HarmitMinhas96 — APPROVED** [review]
Looks good, mostly minor comments.

Additionally, out of scope for this ticket, but maybe we should build some CI around this to run the defined test against some generated dummy models for each model type.

File: clarifai/models/model_serving/pb_model_repository.py [hunk 5]
```diff
@@ -32,24 +32,26 @@ def __init__(self, model_config: Type[TritonModelConfig]):
...
     """
     Write Python Module to file.
 
     Args:
     -----
-    module_name: Python module name to write to file
-    filename: Name of the file to write to destination_dir
-    destination_dir: Directory to save the generated triton model file.
-
+      module_name: Python module name to write to file
```
[the remark this answers is not in the document]
Comment (phatvo9) [reply clarifai/models/model_serving/pb_model_repository.py:41]: Thanks, nice catch!

File: clarifai/models/model_serving/models/model_types.py (hunk 4 above)
Comment (phatvo9) [reply clarifai/models/model_serving/models/model_types.py:42]: I want to split a group of image data into individual images. This way, users can preprocess each image separately. Another reason is to ensure that the data type remains consistent with text input models.
Thank you for suggesting `input_data[:]`, looks much simpler.

File: clarifai/models/model_serving/models/inference.py (hunk 1 above)
Comment (phatvo9) [reply clarifai/models/model_serving/models/inference.py:16]: The point is `inference` imported as module, so if that line is uncommented, an error will occur since `clarifai-model-type` or `MODEL_TYPE_PLACEHOLDER` is not valid model type name. A possible solution is to read `inference` as plain text instead import it, in this case can have more flexible settings.

File: clarifai/models/model_serving/cli/deploy_cli.py
Comment (phatvo9) [reply clarifai/models/model_serving/cli/deploy_cli.py:1]: Included max_batch_size in `clarifai-model-upload-init` and updated doc

File: clarifai/models/model_serving/models/inference.py (hunk 1 above)
Comment (HarmitMinhas96) [reply clarifai/models/model_serving/models/inference.py:16]: One additional option is that `get_model_config` can just log an error/warning when `MODEL_TYPE_PLACEHOLDER` is passed in.

If you prefer the way it currently is, that's ok, but we should definitely add a comment.

File: clarifai/models/model_serving/models/model_types.py (hunk 4 above)
Comment (HarmitMinhas96) [reply clarifai/models/model_serving/models/model_types.py]: But isn't the image data already individual images? I see it like this, but maybe I'm missing something?
```
input_data = [im1, im2, im3]
parse_predictions(input_data):
    a = input_data[:] # a is still [im1, im2, im3]
```

Comment (phatvo9) [reply clarifai/models/model_serving/models/model_types.py]: `input_data` is acquired through `get_input_tensor_by_name()` from triton request, it's numpy array with a shape of [bs, h, w, c].

File: clarifai/models/model_serving/models/inference.py (hunk 1 above)
Comment (phatvo9) [reply clarifai/models/model_serving/models/inference.py]: > One additional option is that get_model_config can just log an error/warning when MODEL_TYPE_PLACEHOLDER is passed in.
Thanks for suggesting, I think this approach is the optimal choice for code clarity and safety.
